### PR TITLE
KEYCLOAK-15633 SSL connection to DB

### PR DESCRIPTION
**Pull request by jonathanvila**
Added new property to pass params to the database : %PREFIX%_CONN_PROPERTY that will be sent to datasource.connection-property in order to allow SSL connection to the DB.
https://issues.redhat.com/browse/KEYCLOAK-15633

Thanks for submitting your Pull Request!

Please make sure your PR meets the following requirements:

- [ ] Pull Request title is properly formatted: `[CLOUD-XYA] Subject`
- [X] Pull Request contains link to the JIRA issue
- [X] Pull Request contains description of the issue
- [X] Pull Request does not include fixes for issues other than the main ticket
- [X] Attached commits represent units of work and are properly formatted
- [ ] You have read and agreed to the Developer Certificate of Origin (DCO) (see `CONTRIBUTING.md`)
- [X] Every commit contains `Signed-off-by: Your Name <yourname@example.com>` - use `git commit -s`

### Steps to Test
1. Build Image
    1. go to the jenkins server : /view/RH-SSO/job/rh-sso-pipeline-continuous_container
    2. check for last execution without "release" in the title
    3. rebuild 
        1. check these critical parameters
            1.  declare -x PIPELINE_EXECUTION_ID="keycloak-15633-sso74"  <-- id 
            2. declare -x IMAGE_OPENSHIFT_REPO="https://github.com/{your repo}/redhat-sso-7-openshift-image.git" <-- your repo with the changes
            3. declare -x IMAGE_OPENSHIFT_BRANCH="KEYCLOAK-15633-allow-ssl-to-db-sso74" <-- branch with changes
            4. declare -x RELEASE="false" <-- **WATCH** specifically this parameter that has to be "false" 
        2. check BUILD_OPENJDK , uncheck BUILD_OPENJ9
        3. clear IMAGE_STANDALONE and IMAGE_OPENSHIFT and IMAGE_OPENSHIFT_OPENJ9
        4. To make it faster ,  ~ 24 minutes instead of several hours
        5. When finished check the Full log and search for "image available as" and **copy the path** on redhat registry
2. Install SSO Operator on Openshift ( using the Operatorhub embedded screen )
    1.  go to the openshift keycloak playground
    2. login as admin
    3. install the SSO operator from the Operatorhub embedded
3. Modify the Operator Yaml ( the CSV ) setting the RELATED_IMAGE_SSO with that image created
    1. + install.spec.deployments.spec.template.spec.containers.env.RELATED_IMAGE_RHSSO= {redhat registry image url copied above}
4. Install CrunchyData Postgres Operator and pgo cli
    1.  https://access.crunchydata.com/documentation/postgres-operator/latest/quickstart/
5. Generate certificates and upload them locally
    1.  https://blog.crunchydata.com/blog/set-up-tls-for-postgresql-in-kubernetes
6. Create a truststore to be used by keycloak
    1.  considering previous steps created `ca.crt` and `server.crt`
    2. keytool -import -file server.crt -alias crunchy-server-cert -keystore truststore.jks
    3. keytool -import -file ca.crt -alias crunchy-ca-cert -keystore truststore.jks
7. Make this trustore available for keycloak
    1. kubectl create configmap sso-truststore-ssl --from-file=truststore.jks -n {your namespace}
8. Create Keycloak CR
    1. in the operator , keycloak tab,  KeycloakDeploymentSpec.Experimental
        1. add Env vars
            1. DB_CONN_PROPERTY=sslMode=require;ssl=true
            2. SCRIPT_DEBUG='true'
            3. KEYCLOAK_POSTGRESQL_SERVICE_HOST=hippo.pgo.svc.cluster.local
            4. DB_USERNAME=testuser
            5. DB_PASSWORD=test
            6. DB_DATABASE=hippo
            7. JAVA_OPTS_APPEND=-Djavax.net.ssl.trustStore=/opt/jboss/.postgresql/truststore.jks    -Djavax.net.ssl.trustStorePassword=nothing -Djavax.net.debug=all
        2. add volumes
            1. Configmap
            2. mount path /opt/jboss/.postgresql
            3. name sso-truststore-ssl
            4. items.key truststore.jks
            5. items.path truststore.jks
9. Check the SSL logs for the Keycloak pod , it should appear some DELETE or SELECT queries inside SSL blocks
```
  0000: 50 00 00 00 35 00 53 45   4C 45 43 54 20 49 44 20  P...5.SELECT ID 
  0010: 46 52 4F 4D 20 74 65 73   74 75 73 65 72 2E 64 61  FROM testuser.da
  0020: 74 61 62 61 73 65 63 68   61 6E 67 65 6C 6F 67 6C  tabasechangelogl
   0030: 6F 63 6B 00 00 00 42 00   00 00 0C 00 00 00 00 00  ock...B.........
   0040: 00 00 00 44 00 00 00 06   50 00 45 00 00 00 09 00  ...D....P.E.....
   0050: 00 00 00 00 53 00 00 00   04 17 00 00 00 00 00 00  ....S...........
   0060: 00 00 00 00 00 00 00 00   00 00                                        ..........```

### CODE REVIEW

**Review by vmuzikar — COMMENTED** [review]
Thank you @jonathanvila, nice work! As far as I can tell LGTM (tested it, works) but I'm no expert in this area so I'd like to wait for a review from @iankko before approving.

We should also think about how to document it. The [RH-SSO image docs](https://access.redhat.com/documentation/en-us/red_hat_single_sign-on/7.4/html/red_hat_single_sign-on_for_openshift_on_openjdk/advanced_concepts#sso-connecting-to-an-external-database) seem to be a bit vague, so I don't think this update belongs there. Maybe I'd create a template for it instead, WDYT @iankko? But that could be possibly done as a follow-up.

File: modules/os-eap-datasource/1.0/added/launch/datasource-common.sh
Comment (iankko) [inline modules/os-eap-datasource/1.0/added/launch/datasource-common.sh]: The `echo` is not needed here. Pass the needed variable to the translate (`tr`) command directly. E.g.:
```
[iankko@localhost ~]$ export conn_properties="first;second;third"
[iankko@localhost ~]$ tr ';' '\n' <<< "${conn_properties}"
first
second
third
```

Comment (iankko) [inline modules/os-eap-datasource/1.0/added/launch/datasource-common.sh]: `properties_array=...` This is not creating an array, but rather string.

To get an array,  either `declare -a properties_array=....`, or enclose the subcommand on the right side of the assignment with another pair of the brackes -- IOW: `properties_array=($(my_subcommand_here arg1 ... argN))`

If the variable doesn't need to be an array, you can keep it as is, but drop the `_array` suffix from it, so it's not confusing.

From the above choose the alternative, which works best for you.

File: modules/os-eap-datasource/1.0/added/launch/datasource-common.sh
Comment (iankko) [inline modules/os-eap-datasource/1.0/added/launch/datasource-common.sh]: If `$properties_array` is to be array, it should be changed to `for element in "${properties_array[@]}"; do`

File: modules/os-eap7-launch/added/launch/datasource.sh
Comment (iankko) [inline modules/os-eap7-launch/added/launch/datasource.sh]: If / Once replacing `driver`, please replace all positional arguments passed to `generate_datasource_common` with their respective / corresponding local names, not just the `driver` one. IOW change it to read like:

```
generate_datasource_common "${pool_name}" ....
```

**Comment by iankko** [comment]
@jonathanvila Thank you for your proposal & sorry again for my delay getting to this.

Couple of (not only administrative related) notes:

* Thank you (extra points!) for the detailed `Steps To Test` (truly appreciated!) Is the operator part needed there? Could those be possibly automated? (just having the RH-SSO image running your new variable & the DB pod support SSL, would it be possible to verify the connection property got appended / is used correctly) If that's the case (the steps can be automated), we could create an automated test as a follow-up for this task / JIRA, WDYT?

Follow-up on `Is the operator part needed there?`:
<hr/>
 Was meant like, of course the change can be verified for work by the operator too. But since the change is done in the RH-SSO for OpenJ{DK,9} for OpenShift images, wondering if we could automate it into a regular test, performing each time a new image build is done.
<hr/>

More notes:
* Please squash all the (`8` so far) commits into one final one.
* The changes LGTM in overall. Just couple of Bash nitpicks to consider (e.g. not to need to launch another external process, when it's not needed, etc.) PTAL at the respective comments above / inline.

**Review by iankko — CHANGES_REQUESTED** [review]
LGTM in general. Just couple of small Bash nitpicks to address yet & squash the intermediary
commits into one final one.

@jonathanvila Great work so far. Thank you for it!

File: modules/os-eap-datasource/1.0/added/launch/datasource-common.sh
Comment (jonathanvila) [reply modules/os-eap-datasource/1.0/added/launch/datasource-common.sh]: :+1 , I see the benefit of your approach on reducing the sub-shells

Comment (jonathanvila) [reply modules/os-eap-datasource/1.0/added/launch/datasource-common.sh]: you are right :+1

File: modules/os-eap7-launch/added/launch/datasource.sh
Comment (jonathanvila) [reply modules/os-eap7-launch/added/launch/datasource.sh]: :+1

**Comment by jonathanvila** [comment]
@iankko why should I squash my commits ? Isn't that what is done with the PR : squash and merge instead of merge.
image

**Comment by iankko** [comment]
> @iankko why should I squash my commits ? Isn't that what is done with the PR : squash and merge instead of merge.
> image

Probably is (re: squash & merge by merging). But so far we are / were used to squash all the commits into one final one (if it's related to single JIRA issue). It's something like a sign, you have finished the work on the feature, and it's considered complete from your PoV.

Can't explain in better way, why squashing is required. Maybe just because it's better for having cleaner history. IDK TBH

@vmuzikar @pskopek Can you think of any other reasoning, why squashing is required?

Thanks, Jan

**Comment by vmuzikar** [comment]
> @vmuzikar @pskopek Can you think of any other reasoning, why squashing is required?

Yes, I see it the same way. Worth noting is also that you should squash the commits after all work is done from your PoV. It's a good practise to address review comments in separate commits (and then squash it once all of them are resolved).

**Comment by jonathanvila** [comment]
I see the point of squashing commits to the upstream project ( that's the function in the screenshot ) but forcing to squash the commits in the fork branch, doesnt seem logic at all for me. What is a requirement for the upstream project should not be a requirement to personal branches in the forks.

To set the PR is ready to be reviewed and finished we can use Github labels, or even the result of the reviewers review.

I dont see why doing a last commit squashing everything in the fork branch ( which eventualy is gonna be deleted ) just as an extra final step , and not only in the upstream is considered any good practice at all.

In the end what the project wants is to have squashed commits features in its code..... not in the temporary and ephemeral personal fork branches. 
Am I right ?

@vmuzikar @pskopek @iankko

**Comment by vmuzikar** [comment]
@jonathanvila I'm afraid this is not the right place for this discussion. Let's bring it to team's ML. ;)

**Comment by jonathanvila** [comment]
@vmuzikar I already sent an email to ML.... but I was following the
discussion here opened by @iankko

**Comment by iankko** [comment]
@jonathanvila  Looking / testing this change more deeply (see previous comment), made me to wonder if we need this code change at all. Looking at the code, there are are `{prefix}_XA_CONNECTION_PROPERTY_*` definitions processed:

https://github.com/jboss-container-images/redhat-sso-7-openshift-image/blob/sso74-dev/modules/os-eap-launch/added/launch/datasource-common.sh#L251

How the change in this PR differs from what's already there? IOW what can be achieved via the new `${prefix_CONN_PROPERTY` what isn't possible via the `${prefix}_XA_CONNECTION_PROPERTY_*` already.

TBH, I have tried to set a new `DB_XA_CONNECTION_PROPERTY_SSLMODE=require` environment variable on RH-SSO pod deployed via `sso74-x509-postgresql-persistent` template, and it failed with error like:
```
...
Caused by: javax.resource.ResourceException: IJ031102: Failed to load XA datasource: org.postgresql.xa.PGXADataSource
--
Caused by: java.lang.NoSuchMethodException: Method setSSLMODE not found
...
```

which seems pretty analogous to `setUSER` method not being available for H2 driver in previous comment.

IOW IMHO the only difference between the existing `${prefix_XA_CONNECTION_PROPERTY_*` functionality, and the new `${prefix}_CONN_PROPERTY` functionality seems the be, the former is doing that in `generate_external_datasource()` routine, while this code change is doing the same setting in `generate_datasource_common()`. Am I missing something?

If that's the case, can you have a look if we really need this change, or the ssl mode configuration to external postgresql server can be done using the existing `${prefix_XA_CONNECTION_PROPERTY_<property_name_here>` functionality? (IOW just documentation change / scenario how to do that is sufficient)

**Comment by iankko** [comment]
Following on previous, what might be potentially needed is to modify the code, not to change the capitalization of the properties passed. Using e.g.:
* `DB_SERVICE_PREFIX_MAPPING=sso-postgresql=DB`
* `DB_XA_CONNECTION_PROPERTY_sslmode=require` and
* `DB_XA_CONNECTION_PROPERTY_ssl=true`
tries to call `setSslmode`, which doesn't seem to be known:
```
Caused by: javax.resource.ResourceException: IJ031102: Failed to load XA datasource: org.postgresql.xa.PGXADataSource
                  at org.jboss.ironjacamar.jdbcadapters@1.4.27.Final-redhat-00001//org.jboss.jca.adapters.jdbc.xa.XAManagedConnectionFactory.getXADataSource(XAManagedConnectionFactory.java:662)
                  at org.jboss.ironjacamar.jdbcadapters@1.4.27.Final-redhat-00001//org.jboss.jca.adapters.jdbc.xa.XAManagedConnectionFactory.getXAManagedConnection(XAManagedConnectionFactory.java:515)
                  ... 56 more
Caused by: java.lang.NoSuchMethodException: Method setSslmode not found
...

**Pull request closed without merging**
